build.ts

```diff
@@ -0,0 +1,381 @@
+  let extraOpts: string[]  = [];
+  for (let i=2; i<argv.length; i++) {
+    const arg = argv[i];
+    if (arg === "--") {
+      extraOpts = [...extraOpts, ...argv.slice(i+1)];
+      break;
+    } else if (arg.startsWith("-")) {
+      const argWithoutDashes = arg.startsWith("--") ? arg.substring(2) : arg.substring(1);
+      switch(argWithoutDashes) {
+        case "production":
+          result.production = true;
+          break;
+        case "e2e":
+          result.e2e = true;
+          break;
+        case "settings":
+          result.settings = argv[++i];
+          break;
+        case "db":
+          result.db = argv[++i];
+          break;
+        case "postgresUrl":
+          result.postgresUrl = argv[++i];
+          break;
+        case "postgresUrlFile":
+          result.postgresUrlFile = argv[++i];
+          break;
+        case "shell":
+          result.shell = true;
+          break;
+        case "command":
+          result.action = "command";
+          result.command = argv[++i];
+          break;
+        case "lint":
+          result.lint = true;
+          break;
+        case "build":
+          result.action = "build";
+          break;
+        case "watch":
+        case "run":
+          result.action = "watch"
+          break;
+        // Ignored arguments for Estrella back-compat
+        case "quiet":
+        case "silent":
+        case "sourcemap":
+        case "inline-sourcemap":
+        case "no-color":
+        case "no-clear":
+        case "no-diag":
+        case "color":
+        case "diag":
+          break;
+      }
+    } else {
+      extraOpts.push(arg);
+    }
+  }
+  
+  return [result,extraOpts];
+}
+
+const [opts, args] = parseCommandLine(process.argv /*, [
+  ["production", "Run in production mode"],
+  ["e2e", "Run in end-to-end testing mode"],
+  ["settings", "A JSON config file for the server", "<file>"],
+  ["db", "A path to a database connection config file", "<file>"],
+  ["postgresUrl", "A postgresql connection connection string", "<url>"],
+  ["postgresUrlFile", "The name of a text file which contains a postgresql URL for the database", "<file>"],
+  ["shell", "Open an interactive shell instead of running a webserver"],
+  ["command", "Run the given server shell command, then exit", "<string>"],
+  ["lint", "Run the linter on site refresh"],
+]*/)
+
+
+const defaultServerPort = 3000;
+
+const getServerPort = () => {
+  if (opts.command) {
+    return 5001;
+  }
+  const port = parseInt(process.env.PORT ?? "");
+  return Number.isNaN(port) ? defaultServerPort : port;
+}
+
+let latestCompletedBuildId: string|null = generateBuildId();
+let inProgressBuildId: string|null = null;
+const serverPort = getServerPort();
+const websocketPort = serverPort + 1;
+
+setOutputDir(`./build${serverPort === defaultServerPort ? "" : serverPort}`);
+
+// Two things this script should do, that it currently doesn't:
+//  * Provide a websocket server for signaling autorefresh
```

Comment:
Is this still missing?  (Also, was there a second thing that it should do that it currently doesn't?)